### 生成数据集对象并分析

In [1]:
from torch_geometric.datasets import Planetoid

如下方代码所示，在PyG中生成一个数据集是简单直接的。在第一次生成PyG内置 的数据集时，程序首先下载原始文件，然后将原始文件处理成包含Data对象的 Dataset对象并保存到文件

In [2]:
datasets = Planetoid(root="../../datasets/Cora", name="Cora")

Processing...
Done!


### 查看数据集

In [6]:
print("数据集图的个数 : {}".format(len(datasets)))
print("数据集分类 : {}".format(datasets.num_classes))  # 7
print("数据集特征 : {}".format(datasets.num_features))  # 1433

数据集图的个数 : 1
数据集分类 : 7
数据集特征 : 1433


如上面代码 可以看到该数据集只有一个图，包含7个分类任务，节点的属性为1433维度

### 分析数据集样本

In [11]:
data = datasets[0]
# 查看图的相关信息
print(data)

print("是否是无向图 : {} ".format(data.is_undirected()))
print("可以用作训练集的节点个数 : {} ".format(data.train_mask.sum().item()))
print("可以用作验证集的节点个数 : {} ".format(data.val_mask.sum().item()))
print("可以用作测试集的节点个数 : {} ".format(data.test_mask.sum().item()))

Data(x=[2708, 1433], edge_index=[2, 10556], y=[2708], train_mask=[2708], val_mask=[2708], test_mask=[2708])
是否是无向图 : True 
可以用作训练集的节点个数 : 140 
可以用作验证集的节点个数 : 500 
可以用作测试集的节点个数 : 1000 


现在我们看到该数据集包含的唯一的图，有2708个节点，节点特征为1433维，有
10556条边，有140个用作训练集的节点，有500个用作验证集的节点，有1000个用 作测试集的节点。

### 数据集的使用

假设我们定义好了一个图神经网络模型，其名为Net。在下方的代码中，我们展示了节点分类图数据集在训练过程中的使用。

In [15]:
import torch
import torch.nn as nn
import torch.nn as nn

class Net(nn.Module):
    def __init__(self):
        pass

    def forward(self, x):
        pass

    def train(self):
        pass

In [14]:
device = "cuda" if torch.cuda.is_available() else "cpu"
# 定义模型, 这个模型仅仅是简单定义, 并不能使用
model = Net().to(device)
# 定义优化函数
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)

model.train()

for epoch in range(200):
    # 重置梯度
    optimizer.zero_grad()
    # 获取模型输出
    out = model(data)
    loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])
    loss.backward()
    optimizer.step()

AttributeError: 'Net' object has no attribute '_modules'